docs/src/index.md

```diff
@@ -4,6 +4,350 @@ DocTestSetup = :(using ChainRulesCore, ChainRules)
+
+The rules are encoded as `rrules` and `frules`, for use in reverse-mode and forward-mode differentiation respectively.
+
+The `rrule` for some function `foo`, which takes the positional argument `args` and keyword argument `kwargs`, is written:
+
+```julia
+function rrule(::typeof(foo), args; kwargs...)
+    ...
+    return y, pullback
+end
+```
+where `y` must be equal to `foo(args; kwargs...)`, and `pullback` is a function to propagate the derivative information backwards at that point (more later).
+
+Similarly, the `frule` is written:
+```julia
+function frule(::typeof(foo), args; kwargs...)
+    ...
+    return y, pushforward
+end
+```
+again `y = foo(args; kwargs...)`, and `pushforward` is a function to propagate the derivative information forwards at that point (more later).
+
+Almost always the _pushforward_/_pullback_ will be declared locally within the `ffrule`/`rrule`, and will be a _closure_ over some of the other arguments.
+
+### The propagators: pushforward and pullback
+
+!!! terminology "pushforward and pullback"
+
+    _Pushforward_ and _pullback_ are fancy words that the autodiff community adopted from Differential Geometry.
+    The are broadly in agreement with the use of [pullback](https://en.wikipedia.org/wiki/Pullback_(differential_geometry)) and [pushforward](https://en.wikipedia.org/wiki/Pushforward_(differential)) in differential geometry.
+    But any geometer will tell you these are the super-boring flat cases. Some will also frown at you.
+    Other terms that may be used include for _pullback_ the **backpropagator**, and by analogy for _pushforward_ the **forwardpropagator**, thus these are the _propagators_.
+    These are also good names because effectively they propagate wiggles and wobbles through them, via the chainrule.
+    (the term **backpropagator** may originate with ["Lambda The Ultimate Backpropagator"](http://www-bcl.cs.may.ie/~barak/papers/toplas-reverse.pdf) by Bearlmutter and Siskind, 2008)
+
+#### Core Idea
+
+ - The **pushforward** takes a wiggle in the _input space_, and tells what wobble you would create in the output space, by passing it through the function.
+ - The **pullback** takes a wobble in the _output space_, and tells you what wiggle you would need to make in the _input_ space to achieve it.
+
+#### The anatomy of pushforward and pullback
+
+For our function `foo(args...; kwargs) = Y`:
+
+The pushforward is a function:
+
+```julia
+function pushforward(Δself, Δargs...)
+    ...
+    return ∂Y
+end
+```
+
+The input to the pushforward is often called the _perturbation_.
+If the function is `y = f(x)` often the pushforward will be written `ẏ = pushforward(ḟ, ẋ)`.
+(`ẏ` is commonly used to represent the permutation for `y`)
+
+!!! note
+
+    There is one `Δarg` per `arg` to the original function.
+    The `Δargs` are similar in type/structure to the corresponding inputs `args` (`Δself` is explained below).
```

Comment:
added
```
### Where are the gradients for keyword arguments?
_pullbacks_ do not return a gradient for keyword arguments;
similarly _pushfowards_ do not accept a pertubation for keyword arguments.
This is because in practice functions are very rarely differentiable with respect to keyword arguments.
As a rule keyword arguments tend to control side-effects, like logging verbsoity,
or to be functionality changing to perform a different operation, e.g. `dims=3`, and thus not differentiable.
To the best of our knowledge no julia AD system, with support for the definition of custom primatives, supports differentating with respect to keyword arguments.
At some point in the future ChainRules may support these. Maybe.
```